### PR TITLE
Container entrypoints

### PR DESCRIPTION
**Pull request by impl**
[Rendered](https://github.com/puppetlabs/relay-rfcs/blob/entrypoints/content/0000-container-entrypoints/rfc.md)

### CODE REVIEW

**Comment by bradhe** [comment]
# Review

Some thoughts on two additional (potential) risks:

1. By injecting code in to our users' environments we're adding another potential attack vector for our users' containers.
1. We're making a lot of assumptions about the types and structure of containers that our users are going to use. I'd be willing to bet this methodology would work with probably 95% of the containers out there in the world right now.

# Next steps

Seems like we can't strictly adopt this RFC until a POC of this approach has been vetted as these RFCs have a way of becoming documentation. Should we just land this RFC for the time being and kick off the POC process, amending the RFC accordingly if the approach changes direction? Or should we leave this as draft 'till we have confirmation on the approach?

**Comment by impl** [comment]
I'll add those risks, thanks!

I think the only part we don't have high certainty of feasibility on is the PVC approach, in which case we can always fall back to using an init container instead (somewhat slower and disliked by the Knative team). I will add this as a contingency, but I'd be comfortable landing it knowing that the mounting strategy might end up slightly different than what we proposed here.

**Review by bradhe — APPROVED** [review]
I'm happy with this approach.

File: content/0000-container-entrypoints/rfc.md [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+  running in two zones instead of three. Since the PVCs will be distributed
+  across tenants we should still retain very high availability guarantees
+  barring other unforeseen issues.
+* There is a cultural risk that the entrypointer binary may feel like trusted
+  code. It is not. It runs directly in a user-controlled environment without any
+  separation from its delegate process. For example, the delegate process might
+  directly read or manipulate the memory of the entrypointer.
+* Conversely, if our own code is insecure, we introduce a potential attack
+  vector to users' containers.
+* We recognize that this may make compatibility with operating systems and
```
Comment (kyleterry) [inline content/0000-container-entrypoints/rfc.md]: We can largely avoid this by using Go, which cross-compiles quite easily to architectures the Linux kernel supports.

**Review by kyleterry — APPROVED** [review]
Looks pretty good to me. I would mention, in the section about untrusted code, that any authentication keys used to communicate with services outside the Pod ~~should~~ MUST be tenant-specific so memory dumps don't leak global credentials (not that we should be using broad service-accounts, but it seems worth mentioning).

**Review by rick-a-lane-ii — APPROVED** [review]

**Pull request merged**